listenbrainz_spark/recommendations/tests/test_models.py

```diff
@@ -22,44 +25,158 @@ def tearDownClass(cls):
+        mock_sqrt.assert_called_once_with(mock_reduce.return_value.__truediv__())
 
     def test_preprocess_data(self):
         test_playcounts_df = utils.read_files_from_HDFS(TEST_PLAYCOUNTS_PATH)
         training_data, validation_data, test_data = train_models.preprocess_data(test_playcounts_df)
         total_playcounts = training_data.count() + validation_data.count() + test_data.count()
         self.assertEqual(total_playcounts, PLAYCOUNTS_COUNT)
 
-    def test_save_model(self):
-        training_data, validation_data, test_data = super().split_playcounts()
+    def test_generate_model_id(self):
+        model_id = train_models.generate_model_id()
+        assert re.match('{}-*'.format(config.MODEL_ID_PREFIX), model_id)
+
+    def test_get_best_model_path(self):
+        model_id = "a36d6fc9-49d0-4789-a7dd-a2b72369ca45"
+        actual_path = train_models.get_best_model_path(model_id)
+        expected_path = config.HDFS_CLUSTER_URI + path.DATA_DIR + '/' + model_id
+        self.assertEqual(actual_path, expected_path)
+
+    def test_get_latest_dataframe_id(self):
+        df_id_1 = "a36d6fc9-49d0-4789-a7dd-a2b72369ca45"
+        df_metadata_dict_1 = self.get_dataframe_metadata(df_id_1)
+        df_1 = utils.create_dataframe(schema.convert_dataframe_metadata_to_row(df_metadata_dict_1),
+                                      schema.dataframe_metadata_schema)
+
+        df_id_2 = "bbbd6fc9-49d0-4789-a7dd-a2b72369ca45"
+        df_metadata_dict_2 = self.get_dataframe_metadata(df_id_2)
+        df_2 = utils.create_dataframe(schema.convert_dataframe_metadata_to_row(df_metadata_dict_2),
+                                      schema.dataframe_metadata_schema)
+
+        df_metadata = df_1.union(df_2)
+        model_metadata = {}
+
+        train_models.get_latest_dataframe_id(df_metadata, model_metadata)
+        self.assertEqual(model_metadata['dataframe_id'], df_id_2)
 
-        best_model, _, best_model_metadata = train_models.train(
-            training_data, validation_data, validation_data.count(), self.ranks,
-            self.lambdas, self.iterations, self.alpha
+    @patch('listenbrainz_spark.recommendations.train_models.train')
+    def test_get_best_model_metadata(self, mock_train):
+        best_model = train_models.Model(
+            model=mock_train.return_value,
+            validation_rmse=3.9,
+            rank=4,
+            lmbda=2.1,
+            iteration=1,
+            model_id="xxx",
+            training_time="3.1",
+            rmse_time="2.1",
+            alpha=3.0,
         )
-        model_save_path = os.path.join('/test/model', best_model_metadata['model_id'])
-        train_models.save_model(model_save_path, best_model_metadata['model_id'], best_model)
-        model_exist = utils.path_exists(model_save_path)
-        self.assertTrue(model_exist)
+
+        metadata = train_models.get_best_model_metadata(best_model)
+        self.assertEqual(best_model.validation_rmse, metadata['validation_rmse'])
```

Comment:
is metadata['validation_rmse'] the same as the constant value of 3.9 from above?

Reply:
yes. we are converting nametuple to a dict in `get_best_model_metadata`